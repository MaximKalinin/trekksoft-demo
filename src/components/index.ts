export { Input } from "@app/components/input";
