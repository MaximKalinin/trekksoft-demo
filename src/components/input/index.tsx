import React, { forwardRef } from "react";
import styles from "@app/components/input/styles.module.less";
import { Spinner } from "@app/components";

interface Props
  extends React.DetailedHTMLProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
  > {
  onSearch?: () => void;
  searchDisabled?: boolean;
  loading?: boolean;
}

export const Input = forwardRef<HTMLInputElement, Props>((props, ref) => {
  const { className, onSearch, searchDisabled, loading, ...other } = props;
  const internalClassName = [className, styles.label].join(" ");
  return (
    <label className={internalClassName}>
      <input {...other} ref={ref} className={styles.label__input} />
      <button
        className={styles.label__button}
        disabled={searchDisabled || loading}
        onClick={onSearch}
      >
        {loading ? <Spinner /> : <span className={styles.button__image} />}
      </button>
    </label>
  );
});
