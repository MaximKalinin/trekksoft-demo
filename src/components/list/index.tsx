import React, { FC, useState, useMemo } from "react";
import styles from "@app/components/list/styles.module.less";
import orderBy from "lodash.orderby";

interface IListProps {
  mainFilter?: {
    text: string;
    property: string;
  };
  secondaryFilter?: {
    text: string;
    property: string;
  };
  items: any[];
  imageProperty?: string;
  loginProperty?: string;
  nameProperty?: string;
  secondaryInfoProperty?: string;
}

type TFilter = "main" | "secondary";

export const List: FC<IListProps> = (props) => {
  const {
    mainFilter,
    secondaryFilter,
    items,
    imageProperty,
    loginProperty,
    nameProperty,
    secondaryInfoProperty,
  } = props;
  const [activeFilter, setActiveFilter] = useState<TFilter>(
    mainFilter ? "main" : "secondary"
  );
  const [order, setOrder] = useState<"asc" | "desc">("asc");
  const handleClick = (filter: TFilter) => () => {
    if (activeFilter === filter) {
      setOrder(order === "asc" ? "desc" : "asc");
      return;
    }
    setActiveFilter(filter);
    setOrder("asc");
  };
  const sortedItems = useMemo(() => {
    const sortProperty =
      activeFilter === "main"
        ? mainFilter?.property
        : secondaryFilter?.property;
    if (!sortProperty) return items;
    return orderBy(items, [sortProperty], [order]);
  }, [activeFilter, order, items]);
  const orderClassName = (filter: TFilter) =>
    `${styles.option__order}${
      (activeFilter === filter && order === "asc") || activeFilter !== filter
        ? ` ${styles["option__order--asc"]}`
        : ""
    }`;
  const optionClassName = (filter: TFilter) =>
    `${styles.filter__option}${
      activeFilter === filter ? ` ${styles["filter__option--active"]}` : ""
    }`;
  console.log(sortedItems, items);
  return (
    <div className={styles.list}>
      <div className={styles.filter}>
        <div className={optionClassName("main")} onClick={handleClick("main")}>
          {mainFilter && (
            <>
              <span>{mainFilter.text}</span>
              <span className={orderClassName("main")} />
            </>
          )}
        </div>
        <div
          className={optionClassName("secondary")}
          onClick={handleClick("secondary")}
        >
          {secondaryFilter && (
            <>
              <span>{secondaryFilter.text}</span>
              <span className={orderClassName("secondary")} />
            </>
          )}
        </div>
      </div>
      <div>
        {sortedItems.map((item, index) => (
          <div className={styles.item} key={index}>
            <img src={item[imageProperty]} />
            <div className={styles.item__main}>
              <span className={styles.main__login}>{item[loginProperty]}</span>
              <span className={styles.main__name}>{item[nameProperty]}</span>
            </div>
            <div className={styles.item__secondary}>
              {item[secondaryInfoProperty]}
            </div>
          </div>
        ))}
      </div>
    </div>
  );
};

List.defaultProps = {
  imageProperty: "avatar_url",
  loginProperty: "login",
  nameProperty: "name",
  secondaryInfoProperty: "repos",
};
