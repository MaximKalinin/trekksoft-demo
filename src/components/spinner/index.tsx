import React, { FC } from "react";

import styles from "@app/components/spinner/styles.module.less";

export const Spinner: FC = () => <div className={styles.loader} />;
