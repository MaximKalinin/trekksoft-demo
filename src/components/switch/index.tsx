import React, { FC, useEffect, useState } from "react";

import styles from "@app/components/switch/styles.module.less";

interface IProps {
  options: Array<{ text: string }>;
  activeOptionId: number;
  onActiveChange: (active: number) => void;
}

export const Switch: FC<IProps> = (props) => {
  const { options, activeOptionId, onActiveChange } = props;
  const [position, setPosition] = useState({ left: 0, width: 0 });
  const setActivePosition = (id: number) => {
    const element = document.querySelector(`[data-option-id="${id}"]`);
    const trackElement = document
      .getElementsByClassName(styles.bottom__track)
      .item(0);
    if (!element || !trackElement) return;
    const { left, width } = element.getBoundingClientRect();
    const { left: trackLeft } = trackElement.getBoundingClientRect();
    setPosition({ left: left - trackLeft, width });
  };
  useEffect(() => {
    setActivePosition(activeOptionId);
  }, [activeOptionId]);
  return (
    <div className={styles.switch}>
      <div className={styles.switch__options}>
        {options.map(({ text }, index) => (
          <div
            className={`${styles.switch__option}${
              index === activeOptionId
                ? ` ${styles["switch__option--active"]}`
                : ""
            }`}
            key={index}
            data-option-id={index}
            onClick={() => onActiveChange(index)}
          >
            <span className={styles.option__text}>{text}</span>
          </div>
        ))}
      </div>
      <div className={styles.bottom__track}>
        <span className={styles.bottom__active} style={{ ...position }}></span>
      </div>
    </div>
  );
};
