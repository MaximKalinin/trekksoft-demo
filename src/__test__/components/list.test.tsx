import React from "react";
import TestRenderer from "react-test-renderer";
import { List } from "@app/components";

describe("List component", () => {
  it("should match snapshot", () => {
    const list = TestRenderer.create(<List items={[]} />);
    expect(list).toMatchSnapshot();
  });
});
