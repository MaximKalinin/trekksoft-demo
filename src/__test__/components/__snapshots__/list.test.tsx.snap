// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should match snapshot 1`] = `
<div
  className="list"
>
  <div
    className="filter"
  >
    <div
      className="filter__option"
      onClick={[Function]}
    />
    <div
      className="filter__option filter__option--active"
      onClick={[Function]}
    />
  </div>
  <div />
</div>
`;
